app/services/action_create_service.rb

```diff
@@ -83,13 +83,18 @@ def deny(comments)
 
   def cancel(comments)
     raise Exceptions::InvalidStateTransitionError, "Only root level request can be canceled" unless request.root?
-
-    if Request::FINISHED_STATES.include?(request.state)
-      raise Exceptions::InvalidStateTransitionError, "The request has already finished"
-    end
+    raise Exceptions::InvalidStateTransitionError, "The request has already finished" if request.finished?
 
     {:state => Request::CANCELED_STATE, :decision => Request::CANCELED_STATUS}.tap do |h|
       h[:reason] = comments if comments
     end
   end
+
+  def error(comments)
+    raise Exceptions::InvalidStateTransitionError, "Current request has already finished" if request.finished?
+    raise Exceptions::ApprovalError, "Failure reason is missing" unless comments
```

Comment:
@bzwei Should this be listed here
https://github.com/RedHatInsights/approval-api/blob/c5e99a5c5ad38597748309e872a7190df8247971/config/initializers/custom_exception_mappings.rb#L8
Also are Approval Errors bad request or server errors. A bad request would mean that the user entered something bad in the dialogs but a server internal error would mean the users request is OK but something is broken in the back end. A server error 5xx can trigger some cleanup activity on the part of the Administrator, who might have to fix his Approval workflow e.g. remove some stale groups.